File: src/main/scala/no/bekk/scala/exercises/Ex6_Computer.scala
package no.bekk.scala.exercises

object Ex6_Computers {

	// A computer with Intel CPU, Asus motherboard, and Corsair memory
	def computer1: Computer = (new ComputerContext with IntelCpuContext with AsusMotherboardContext with CorsairMemoryContext).computer

	// A computer with AMD CPU, MSI motherboard and Kingston memory
	def computer2: Computer = (new ComputerContext with AmdCpuContext with MsiMotherboardContext with KingstonMemoryContext).computer
}

trait CpuContext {
	val cpu: Cpu
}

trait MemoryContext {
	val memory: Memory
}

trait MotherboardContext {
	val motherboard: Motherboard
}

trait ComputerContext {
	self: MotherboardContext with CpuContext with MemoryContext =>
	lazy val computer = Computer(motherboard, cpu, memory)
}

/**
 * Motherboard contexts
 */

trait AsusMotherboardContext extends MotherboardContext {
	lazy val motherboard = Motherboard("Asus")
}

trait MsiMotherboardContext extends MotherboardContext {
	lazy val motherboard = Motherboard("MSI")
}

/**
 * CPU contexts
 */

trait IntelCpuContext extends CpuContext {
	lazy val cpu = Cpu("Intel", 3.5)
}

trait AmdCpuContext extends CpuContext {
	lazy val cpu = Cpu("AMD", 2.8)
}

/**
 * Memory contexts
 */

trait CorsairMemoryContext extends MemoryContext {
	lazy val memory = Memory("Corsair", 16)
}

trait KingstonMemoryContext extends MemoryContext {
	lazy val memory = Memory("Kingston", 8)
}

/**
 * Classes
 */

trait ComputerComponent {
	val manufacturer: String
}
case class Cpu(manufacturer: String, clockSpeed: Double) extends ComputerComponent
case class Memory(manufacturer: String, ram: Int) extends ComputerComponent
case class Motherboard(manufacturer: String) extends ComputerComponent

case class Computer(motherboard: Motherboard, cpu: Cpu, memory: Memory)

File: src/main/java/no/bekk/java/exercises/Ex6_Computer.java
package no.bekk.java.exercises;

public class Ex6_Computer {

	static Computer computer1() {
		// A computer with Intel CPU, Asus motherboard, and Corsair memory
		class Computer1Context implements ComputerContext, IntelCpuContext, AsusMotherboardContext, CorsairMemoryContext {}
		return new Computer1Context().getComputer();
	}

	static Computer computer2() {
		// A computer with AMD CPU, MSI motherboard and Kingston memory
		class Computer2Context implements ComputerContext, AmdCpuContext, MsiMotherboardContext, KingstonMemoryContext {}
		return new Computer2Context().getComputer();
	}
}

interface CpuContext {
	Cpu getCpu();
}

interface MemoryContext {
	Memory getMemory();
}

interface MotherboardContext {
	Motherboard getMotherboard();
}

interface ComputerContext extends CpuContext, MemoryContext, MotherboardContext {

	default Computer getComputer() {
		return new Computer(getMotherboard(), getCpu(), getMemory());
	}
}

/**
 * Motherboard contexts
 */

interface AsusMotherboardContext extends MotherboardContext {
	@Override
	default Motherboard getMotherboard() {
		return new Motherboard("Asus");
	}
}

interface MsiMotherboardContext extends MotherboardContext {
	@Override
	default Motherboard getMotherboard() {
		return new Motherboard("MSI");
	}
}

/**
 * CPU contexts
 */

interface IntelCpuContext extends CpuContext {
	@Override
	default Cpu getCpu() {
		return new Cpu("Intel", 3.5);
	}
}

interface AmdCpuContext extends CpuContext {
	@Override
	default Cpu getCpu() {
		return new Cpu("AMD", 2.8);
	}
}

/**
 * Memory contexts
 */

interface CorsairMemoryContext extends MemoryContext {
	@Override
	default Memory getMemory() {
		return new Memory("Corsair", 16);
	}
}

interface KingstonMemoryContext extends MemoryContext {
	@Override
	default Memory getMemory() {
		return new Memory("Kingston", 8);
	}
}

/**
 * Classes
 */

interface ComputerComponent {
	String getManifacturer();
}

class Cpu implements ComputerComponent {
	public final String manifacturer;
	public final Double clockSpeed;

	public Cpu(String manifacturer, Double clockSpeed) {
		this.manifacturer = manifacturer;
		this.clockSpeed = clockSpeed;
	}

	public String getManifacturer() {
		return manifacturer;
	}
}

class Memory implements ComputerComponent {
	public final String manifacturer;
	public final Integer ram;

	public Memory(String manifacturer, Integer ram) {
		this.manifacturer = manifacturer;
		this.ram = ram;
	}

	public String getManifacturer() {
		return manifacturer;
	}
}

class Motherboard implements ComputerComponent {
	public final String manifacturer;

	public Motherboard(String manifacturer) {
		this.manifacturer = manifacturer;
	}

	public String getManifacturer() {
		return manifacturer;
	}
}

class Computer {
	public final Motherboard motherboard;
	public final Cpu cpu;
	public final Memory memory;

	public Computer(Motherboard motherboard, Cpu cpu, Memory memory) {
		this.motherboard = motherboard;
		this.cpu = cpu;
		this.memory = memory;
	}
}